cmd/soroban-cli/src/commands/contract/build.rs

```diff
@@ -163,23 +168,43 @@ impl Cmd {
             .map(|f| f.split(&[',', ' ']).map(String::from).collect())
     }
 
-    fn packages(&self, metadata: &Metadata) -> Vec<Package> {
-        metadata
+    fn packages(&self, metadata: &Metadata) -> Result<Vec<Package>, Error> {
+        // Filter by the package name if one is provided, or by the package that
+        // matches the manifest path if the manifest path matches a specific
+        // package.
+        let name = if let Some(name) = self.package.clone() {
+            Some(name)
+        } else {
+            // When matching a package based on the manifest path, match against the
+            // absolute path because the paths in the metadata are absolute. Match
+            // against a manifest in the current working directory if no manifest is
+            // specified.
+            let manifest_path = path::absolute(&self.manifest_path).map_err(Error::AbsolutePath)?;
+            metadata
+                .packages
+                .iter()
+                .find(|p| p.manifest_path == manifest_path)
+                .map(|p| p.name.clone())
+        };
+
+        let package = metadata
             .packages
             .iter()
             .filter(|p|
-                // Filter by the package name if one is provided.
-                self.package.is_none() || Some(&p.name) == self.package.as_ref())
-            .filter(|p| {
-                // Filter crates by those that build to cdylib (wasm), unless a
-                // package is provided.
-                self.package.is_some()
-                    || p.targets
+                // Filter by the package name if one is selected based on the above logic.
+                if let Some(name) = &name {
+                    &p.name == name
+                } else {
+                    // Otherwise filter crates by those that build to cdylib (wasm).
+                    p.targets
                         .iter()
                         .any(|t| t.crate_types.iter().any(|c| c == "cdylib"))
-            })
+                }
+            )
             .cloned()
-            .collect()
+            .collect();
+
+        Ok(package)
```

Comment:
Nomenclature nit: if this is a multi-element vector, I think it would be named something plural? I am actually not sure if it's a single-element vector or multi.